README.md
# Numpad

an numeric Keyboard Layout supports english and arabic digits

## Features

- Documented:** each property has a code documentation
- Languages:** supports English and Arabic Digits
- Flexibility:** smooth and flexible to modify anything

# Usage

| Parameter             |            Default             |                                                Description |
|-----------------------|:------------------------------:|-----------------------------------------------------------:|
| **onType**            |            required            |                             Callback when number pressed . |
| **mainAxisAlignment** | MainAxisAlignment.spaceBetween |                                   for spacing horizontal . |
| **padding**           |        horizontal = 30         |                            padding over the whole layout . |
| **leftWidget**        |              null              |                          the widget at the left of the 0 . |
| **rightWidget**       |              null              |                         the widget at the right of the 0 . |
| **highlightColor**    |       Color(0xFFC9C9C9)        |                               on Holding press any number. |
| **runSpace**          |               40               |                                         spacing vertical . |
| **numberStyle**       |              null              |                                        custom number style |
| **radius**            |               45               |                       radius for the shape of the number . |
| **arabicDigits**      |             false              |                       Displays arabic digits if equal true |
| **returnItAsEnglish** |             false              | it returns the digits as english even it's arabic digits . |

#### Example
```dart
 NumPad(
   arabicDigits: true,
   onType: (value) {
     number += value;
     setState(() {});
   },
   rightWidget: IconButton(
   icon: const Icon(Icons.backspace),
   onPressed: () {
   if (number.isNotEmpty) {
     number = number.substring(0, number.length - 1);
      setState(() {});
   }
   },
  ),
);
```

<img src="https://user-images.githubusercontent.com/55635328/216777871-0480f38b-b3a0-4c24-9b6e-1b119e17a3f2.png" width="400" height="600">

also you can find The sample code under the `example` package.

CHANGELOG.md
## 0.0.1

### First Version

- Documented Code
- Arabic Digits
- flexibility to return it as english or arabic
- flexible styles for the layout and Digits
## 0.0.3

### First Version

- Bug Fixes

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:numpad_layout/widgets/numPad.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String number = "";

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'NumPad Example',
      home: Scaffold(
        body: Padding(
          padding: const EdgeInsets.only(top: 150),
          child: Center(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Text(
                  number,
                  style: const TextStyle(
                    fontSize: 32,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(
                  height: 40,
                ),
                NumPad(
                  arabicDigits: true,
                  onType: (value) {
                    number += value;
                    setState(() {});
                  },
                  rightWidget: IconButton(
                    icon: const Icon(Icons.backspace),
                    onPressed: () {
                      if (number.isNotEmpty) {
                        number = number.substring(0, number.length - 1);
                        setState(() {});
                      }
                    },
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}


lib/extension/numbers.dart
extension Numbers on String {
  String get toArabicNumbers {
    Map numbers = {
      '0': '\u0660',
      '1': '\u0661',
      '2': '\u0662',
      '3': '\u0663',
      '4': '\u0664',
      '5': '\u0665',
      '6': '\u0666',
      '7': '\u0667',
      '8': '\u0668',
      '9': '\u0669',
    };
    String number = this;
    numbers.forEach((key, value) => number = number.replaceAll(key, value));
    return number;
  }

  String get toEnglishNumbers {
    Map numbers = {
      '\u0660': '0',
      '\u0661': '1',
      '\u0662': '2',
      '\u0663': '3',
      '\u0664': '4',
      '\u0665': '5',
      '\u0666': '6',
      '\u0667': '7',
      '\u0668': '8',
      '\u0669': '9',
    };
    String number = this;
    numbers.forEach((key, value) => number = number.replaceAll(key, value));
    return number;
  }
}


lib/widgets/num_button.dart
import 'package:flutter/material.dart';
import 'package:numpad_layout/extension/numbers.dart';

class NumButton extends StatelessWidget {
  final String number;
  final VoidCallback onTap;
  final Color? highlightColor;
  final TextStyle? numberStyle;
  final double? radius;
  final bool? arabicDigits;

  const NumButton({
    Key? key,
    required this.number,
    required this.onTap,
    this.highlightColor,
    this.numberStyle,
    this.radius,
    this.arabicDigits = false,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return InkWell(
      borderRadius: BorderRadius.circular(radius ?? 45),
      onTap: onTap,
      highlightColor: highlightColor ?? const Color(0xFFC9C9C9),
      child: Container(
        alignment: Alignment.center,
        width: 50,
        height: 50,
        child: Text(
          arabicDigits == true
              ? number.toArabicNumbers
              : number.toEnglishNumbers,
          style: numberStyle ??
              const TextStyle(
                color: Colors.black,
                fontSize: 32,
                fontWeight: FontWeight.w700,
              ),
        ),
      ),
    );
  }
}


lib/widgets/numpad.dart
import 'package:flutter/material.dart';
import 'package:numpad_layout/extension/numbers.dart';
import 'package:numpad_layout/widgets/num_button.dart';

class NumPad extends StatelessWidget {
  /// Callback when number pressed
  final void Function(String number) onType;

  /// for spacing horizontal default [MainAxisAlignment.spaceBetween]
  final MainAxisAlignment mainAxisAlignment;

  // padding over the whole layout default [horizontal = 30]
  final EdgeInsets? padding;

  /// the widget at the left of the 0
  final Widget? leftWidget;

  /// the widget at the right of the 0
  final Widget? rightWidget;

  /// on Holding pressed any number default [Color(0xFFC9C9C9)]
  final Color? highlightColor;

  /// spacing vertical default [40]
  final double runSpace;

  /// custom number style
  final TextStyle? numberStyle;

  /// radius for the shape of the number default [45]
  final double? radius;

  /// Displays arabic digits if equal [true] default is [false] = english
  final bool? arabicDigits;

  /// it returns the digits as english even it's arabic digits default [false] it return it what it is
  final bool returnItAsEnglish;

  const NumPad({
    Key? key,
    required this.onType,
    this.mainAxisAlignment = MainAxisAlignment.spaceBetween,
    this.padding,
    this.leftWidget,
    this.rightWidget,
    this.highlightColor,
    this.runSpace = 40,
    this.numberStyle,
    this.radius,
    this.arabicDigits,
    this.returnItAsEnglish = false,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    void onTypeNumber(int number) {
      String num = arabicDigits == true ? number.toString().toArabicNumbers : number.toString();

      onType(returnItAsEnglish == true ? num.toEnglishNumbers : num);
    }

    return Padding(
      padding: padding ?? const EdgeInsets.symmetric(horizontal: 30),
      child: Column(
        children: [
          Row(
            mainAxisAlignment: mainAxisAlignment,
            children: [
              for (int number = 1; number <= 3; number++)
                NumButton(
                  number: number.toString(),
                  highlightColor: highlightColor,
                  numberStyle: numberStyle,
                  radius: radius,
                  arabicDigits: arabicDigits,
                  onTap: () => onTypeNumber(number),
                ),
            ],
          ),
          SizedBox(
            height: runSpace,
          ),
          Row(
            mainAxisAlignment: mainAxisAlignment,
            children: [
              for (int number = 4; number <= 6; number++)
                NumButton(
                  number: number.toString(),
                  highlightColor: highlightColor,
                  numberStyle: numberStyle,
                  radius: radius,
                  arabicDigits: arabicDigits,
                  onTap: () => onTypeNumber(
                    number,
                  ),
                ),
            ],
          ),
          SizedBox(
            height: runSpace,
          ),
          Row(
            mainAxisAlignment: mainAxisAlignment,
            children: [
              for (int number = 7; number <= 9; number++)
                NumButton(
                  number: number.toString(),
                  highlightColor: highlightColor,
                  numberStyle: numberStyle,
                  radius: radius,
                  arabicDigits: arabicDigits,
                  onTap: () => onTypeNumber(number),
                ),
            ],
          ),
          SizedBox(
            height: runSpace,
          ),
          Row(
            mainAxisAlignment: mainAxisAlignment,
            children: [
              SizedBox(
                height: 50,
                width: 50,
                child: leftWidget ?? const SizedBox(),
              ),
              NumButton(
                number: "0",
                highlightColor: highlightColor,
                numberStyle: numberStyle,
                radius: radius,
                arabicDigits: arabicDigits,
                onTap: () => onTypeNumber(0),
              ),
              SizedBox(
                width: 50,
                height: 50,
                child: rightWidget ?? const SizedBox(),
              )
            ],
          ),
        ],
      ),
    );
  }
}
